crates/provider/src/new.rs

```diff
@@ -37,18 +43,64 @@ pub struct RootProvider<N, T> {
     pub(crate) inner: Arc<RootProviderInner<N, T>>,
 }
 
+impl<N, T> Clone for RootProvider<N, T> {
+    fn clone(&self) -> Self {
+        Self { inner: self.inner.clone() }
+    }
+}
+
+impl<N, T: fmt::Debug> fmt::Debug for RootProvider<N, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RootProvider").field("client", &self.inner.client).finish_non_exhaustive()
+    }
+}
+
 impl<N: Network, T: Transport> RootProvider<N, T> {
-    /// Create a new root provider.
+    /// Creates a new root provider from the given RPC client.
     pub fn new(client: RpcClient<T>) -> Self {
         Self { inner: Arc::new(RootProviderInner::new(client)) }
     }
 }
 
 impl<N: Network, T: Transport + Clone> RootProvider<N, T> {
-    async fn new_pending_transaction(&self, tx_hash: B256) -> TransportResult<PendingTransaction> {
-        // TODO: Make this configurable.
-        let cfg = WatchConfig::new(tx_hash);
-        self.get_heart().watch_tx(cfg).await.map_err(|_| TransportErrorKind::backend_gone())
+    /// Boxes the inner client.
+    ///
+    /// This will create a new provider if this instance is not the only reference to the inner
+    /// client.
+    pub fn boxed(self) -> RootProvider<N, BoxTransport> {
+        let inner = Arc::unwrap_or_clone(self.inner);
+        RootProvider { inner: Arc::new(inner.boxed()) }
+    }
+
+    /// Gets the subscription corresponding to the given RPC subscription ID.
+    #[cfg(feature = "pubsub")]
+    pub async fn get_subscription<R: RpcReturn>(
+        &self,
+        id: U256,
+    ) -> TransportResult<Subscription<R>> {
+        self.pubsub_frontend()?.get_subscription(id).await.map(Subscription::from)
+    }
+
+    /// Unsubscribes from the subscription corresponding to the given RPC subscription ID.
+    #[cfg(feature = "pubsub")]
+    pub fn unsubscribe(&self, id: U256) -> TransportResult<()> {
+        self.pubsub_frontend()?.unsubscribe(id)
+    }
+
+    #[cfg(feature = "pubsub")]
+    fn pubsub_frontend(&self) -> TransportResult<&PubSubFrontend> {
+        let t = self.transport() as &dyn std::any::Any;
+        t.downcast_ref::<PubSubFrontend>()
+            .or_else(|| {
+                t.downcast_ref::<BoxTransport>()
+                    .and_then(|t| t.as_any().downcast_ref::<PubSubFrontend>())
+            })
+            .ok_or_else(TransportErrorKind::pubsub_unavailable)
+    }
```

Comment:
I am OK with this -- @mattsse @onbjerg any objections?

Reply:
we can probably improve this, but for now this seems fine